File: leetcode/307-range-sum-query-mutable/python/num_array.py
class NumArray:
    
    def __init__(self, nums: list[int]):
        total = 0
        for i in range(len(nums)):
            nums[i] += total
            total = nums[i]
        self.nums = nums
    
    
    def update(self, index: int, val: int) -> None:
        diff = val - (self.nums[index] if index == 0 else self.nums[index] - self.nums[index - 1])
        for i in range(index, len(self.nums)):
            self.nums[i] += diff

    def sumRange(self, left: int, right: int) -> int:
        if left == 0:
            return self.nums[right]
        return self.nums[right] - self.nums[left - 1]


if __name__ == '__main__':
    numArr = NumArray([-2, 0, 3, -5, 2, -1])
    print(numArr.sumRange(0, 2))
    print(numArr.sumRange(2, 5))
    print(numArr.sumRange(0, 5))

File: leetcode/307-range-sum-query-mutable/java/NumArray.java
public class NumArray {

    public static void main(String[] args) {
        int[] arr = {-2, 0, 3, -5, 2, -1};
        NumArray numArray = new NumArray(arr);
        System.out.println(numArray.sumRange(0, 2));
        System.out.println(numArray.sumRange(2, 5));
        System.out.println(numArray.sumRange(0, 5));
    }

    private int[] nums;
    
    public NumArray(int[] nums) {
        int total = 0;
        for (int i = 0; i < nums.length; i += 1) {
            nums[i] += total;
            total = nums[i];
        }
        this.nums = nums;
    }

    public void update(int index, int val) {
        int diff = val - (index == 0 ? nums[index] : nums[index] - nums[index - 1]);
        for (int i = index; i < nums.length; i += 1) {
            nums[i] += diff;
        }
    }
    
    public int sumRange(int left, int right) {
        if (left == 0)  return nums[right];
        return nums[right] - nums[left - 1];
    }
}